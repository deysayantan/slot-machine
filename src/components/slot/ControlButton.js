import React, { Component } from 'react'

class ControlButton extends Component {
    constructor(props) {
        super(props)
        this.state = {
            started : false,
            label : "Start"
        }
    }
    componentDidUpdate(prevProps,prevStates){
        if(prevProps.slotStatus !== this.props.slotStatus){
            this.setState({
                label:this.props.slotStatus ? "Stop": "Start",
                started:this.props.slotStatus
            })
        }
    }
    changeButtonState = () =>{
        this.setState(prevState => {
            return {
                started : !prevState.started,
                label: !prevState.started ? "Stop" : "Start"
            }
        },()=>{
            this.props.userInteractionCallback(this.state.started)
        })
    }
    render() {
        return (
            <button onClick={this.changeButtonState}>
                {this.state.label}
            </button>
        )
    }
}

export default ControlButton
