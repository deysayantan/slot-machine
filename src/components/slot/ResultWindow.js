import React, { Component } from 'react'
import {connect} from 'react-redux'
class ResultWindow extends Component {
    constructor(props) {
        super(props)
        this.state = {
            shouldAppear : false,
            endingCombination:[]
        }
        this.timer = null
    }  
    componentDidUpdate(prevProps,prevStates){
        if(prevProps.slotRunning.slotRunning !== this.props.slotRunning.slotRunning){
            if(this.props.slotRunning.slotRunning){
                clearTimeout(this.timer)
                this.setShouldAppear(false)
            }
            if(!this.props.slotRunning.slotRunning && this.props.userInteraction.userActed){
                clearTimeout(this.timer)
                this.setShouldAppear(true)
                const endingCombination = this.props.result.result
                this.setState({endingCombination})
                this.timer = setTimeout(()=>{
                   this.setShouldAppear(false)
                }, 3000);
            }
        }
    }
    setShouldAppear = (flag) => {
        this.setState({
            shouldAppear : flag
        })
    }
    render() {
        return (
            <div>
                {this.state.shouldAppear && <div>Prize {JSON.stringify(this.state.endingCombination)}</div>}
            </div>
        )
    }
}


const mapStateToProps = state => state;


export default connect(
    mapStateToProps,
    null
)(ResultWindow);
