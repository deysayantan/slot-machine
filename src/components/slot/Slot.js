import React, { Component } from 'react'
import { connect } from 'react-redux';
import Banana from '../../Assets/banana.jpg'
import Orange from '../../Assets/orange.jpg'
import Strawberry from '../../Assets/strawberry.jpg'
import Monkey from '../../Assets/monkey.jpg'
import { Styles } from '../styles/AppStyle';
import {saveResult} from '../../action'

const options = [1,2,3,4]
class Slot extends Component {
    constructor(props) {
        super(props)
        this.state = {
            label : this.sample(options),
            chosenImage : null
        }
    }
    componentDidMount(){
        this.chooseImage()
        this.startInterval(5000)
    }
    componentDidUpdate(prevProps,prevStates){
        if(prevProps.userInteraction !== this.props.userInteraction){
            clearInterval(this.interval)
            console.log(`${this.props.serial} ::: ${this.state.label}`)
            if(this.props.userInteraction){
                this.startInterval(50)
                setTimeout(()=> {
                    clearInterval(this.interval)
                    const payload = {
                        "key":this.props.serial,
                        "value":this.state.label
                    }
                    this.props.saveResult(payload)
                },10000) 
            }
            else{
                this.startInterval(5000)
                console.log("called")
            }
            
        }
    }
    componentWillUnmount(){
        clearInterval(this.interval)
    }
    startInterval = (msec) => {
        this.interval = setInterval(
            () => {
                this.setState({
                    label:this.sample(options)
                },()=>{
                    this.chooseImage()
                })
            }
        ,msec)
    }
    sample = arr => arr[Math.floor(Math.random() * arr.length)];
    chooseImage = () => {
        let image;
        switch(this.state.label){
            case 1:
                image = Banana
                break;
            case 2:
                image = Orange
                break;
            case 3:
                image = Monkey
                break;
            case 4:
                image = Strawberry
                break;
            default:
                image = null
                break;
        }
        this.setState({
            chosenImage: image
        })
    }
    render() {
        console.log(this.props)
        return (
            <img src={this.state.chosenImage} alt="icon" style={Styles.image} />
        )
    }
}

const mapStateToProps = ({ result }) => ({
    result,
});

const mapDispatchToProps = dispatch => ({
    saveResult: (payload) => dispatch(saveResult(payload)),
});

export default connect(
    mapStateToProps,
    mapDispatchToProps,
)(Slot);


