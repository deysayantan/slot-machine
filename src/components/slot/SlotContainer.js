import React, { Component } from 'react'
import { SlotContainerStyles } from '../styles/AppStyle'
import Slot from './Slot'
import ControlButton from './ControlButton'
import {connect} from 'react-redux'
import {setUserInteraction,setSlotRunningStatus} from '../../action'
import ResultWindow from './ResultWindow'
class SlotContainer extends Component {
    constructor(props) {
        super(props)
        this.state = {
            startSlot : false,
        }
        this.timer = null
    }
    componentDidMount(){
        const payload = {
            "userActed" : false
        }
        this.props.setSlotRunningStatus(false)
        this.props.setUserInteraction(payload)
        this.timer = setTimeout(()=>{
            this.startStopSlot(true)
        }, 5000);
    }
    updateUserInteraction = (flag) => {
        this.startStopSlot(flag)
        clearTimeout(this.timer)
        if(!flag){
            this.timer = setTimeout(()=>{
                const payload = {
                    "userActed" : false
                }
                this.props.setUserInteraction(payload)
                this.startStopSlot(true)
            }, 5000);
        }
    }
    resetSlotMachine = () =>{
        clearTimeout(this.timer)
        this.startStopSlot(false)
        this.timer = setTimeout(()=>{
            const payload = {
                "userActed" : false
            }
            this.props.setUserInteraction(payload)
            this.startStopSlot(true)
        }, 5000);
    }
    startStopSlot = (flag) =>{
        this.props.setSlotRunningStatus(flag)
        this.setState({
            startSlot : flag
        })
    }
    render() {
        return (
            <div style={SlotContainerStyles.container}>
                <Slot 
                    resetSlotMachine={this.resetSlotMachine}
                    userInteraction={this.state.startSlot} 
                />
                <ControlButton 
                    slotStatus={this.state.startSlot}
                    userInteractionCallback={this.updateUserInteraction} 
                />
                <ResultWindow />
            </div>
        )
    }
}

const mapStateToProps = ({ userInteraction,slotRunning }) => ({
    userInteraction,
    slotRunning
});

const mapDispatchToProps = dispatch => ({
    setUserInteraction: (payload) => dispatch(setUserInteraction(payload)),
    setSlotRunningStatus : flag => dispatch(setSlotRunningStatus(flag))
});

export default connect(
    mapStateToProps,
    mapDispatchToProps,
)(SlotContainer);
