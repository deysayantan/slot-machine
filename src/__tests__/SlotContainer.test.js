import React from 'react';
import ReactDOM from 'react-dom';
import SlotContainerElement from '../components/slot/SlotContainer';
import Adapter from 'enzyme-adapter-react-16'
import { configure,shallow,mount } from 'enzyme'
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import {SlotContainer} from '../components/slot/SlotContainer'

const mockStore = configureMockStore();
const store = mockStore({});
const mockCallBack = jest.fn(); 
configure({ adapter: new Adapter() })

const slotContainerElement = shallow((<Provider store={store}><SlotContainerElement /></Provider>));

describe('Slot container snapshot testing',()=>{
  it('renders app module correctly',()=>{
      expect(slotContainerElement).toMatchSnapshot();
  });
});


describe('Check slot spinning timings',()=>{
  const wrapper = shallow(
    <Provider store={store}>
      <SlotContainerElement />
    </Provider>
  )
  const componentInstance = wrapper.instance();
  const slotContainerElement = new componentInstance.props.children.type.WrappedComponent(0)
  
  it('Slots should not run immediately after starting',()=>{
    expect(slotContainerElement.state.startSlot).toEqual(false)
  })
})



