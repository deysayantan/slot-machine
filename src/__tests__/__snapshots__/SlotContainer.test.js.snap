// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slot container snapshot testing renders app module correctly 1`] = `ShallowWrapper {}`;
